File: 0034 Find First and Last Position of Element in Sorted Array/find_first_and_last_position_of_element_in_sorted_array.cpp
/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
public:
    int lowerBound(vector<int>& nums, int target, int low, int high) {
        int mid;
        while(low<=high) {
            mid = low + (high-low)/2;
            if(nums[mid]<target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return low;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans(2, -1);
        if(nums.size()==0) return ans;
        int firstOcr = lowerBound(nums, target, 0, nums.size()-1);
        int lastOcr = lowerBound(nums, target+1, 0, nums.size()-1)-1;
        if(firstOcr < nums.size() && nums[firstOcr]==target) {
            return {firstOcr, lastOcr};
        }
        return ans;
    }
};

/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int n = nums.length;
        int[] ans = new int[2];
        ans[0] = -1;
        ans[1] = -1;
        if(n==0) return ans;
        int st = 0;
        int end = n-1;
        int mid = 0;
        while(st<=end) {
            mid = st + (end-st)/2;
            if(nums[mid]==target) {
                for(int i = 0; i<n; i++) {
                    if(nums[i]==target && ans[0]==-1) {
                        ans[0] = i;
                    }
                    if(nums[i]==target) {
                        ans[1] = i;
                    }
                }
                return ans;
            }
            if(nums[mid]<target) {
                st = mid+1;
            } else {
                end = mid-1;
            }
        }
        return ans;
    }
}

/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/
